arraycontext/context.py

```diff
@@ -308,7 +313,7 @@ def call_loopy(self,
         """
 
     @abstractmethod
-    def freeze(self, array: Array) -> Array:
+    def freeze(self, array: ArrayOrContainerT) -> ArrayOrContainerT:
```

Comment:
I'm not sure this is accurate, e.g. `pytato` can take a `pt.Array` and return a `cl.Array` and this constraints the input and output to have the same type.

Maybe something like 
```suggestion
    def freeze(self, array: Union[Array, ContainerT, _ScalarLike]) -> Union[Array, ContainerT, _ScalarLike]:
```
? I'm not sure that works though.. may need to [overload](https://docs.python.org/3/library/typing.html#typing.overload) to get this vaguely accurate..

Typing is hard :(

Reply:
Thanks @alexfikl!   I am not sure what to do with this one.  I think I understand the concern - but not sure how to resolve it properly.  I bent your suggestion just a bit to be consistent with @inducer's comments on #180.

Reply:
Yeah, I'm not sure it's worth diving too much into this until `mypy` starts complaining or we want to use `--strict`. For now, a "good enough" annotation and some clear docs should do the trick!

Reply:
I agree with @alexfikl that typing is hard. And I think arraycontext is sufficiently reliant on Python's dynamic typing that trying to get it fully statically typed is probably a fool's errand. That said, e386416 is my attempt at cleaning things up a bit. I'd be grateful if the both of you could take a look.